src/pipeline_migration/migrate.py

```diff
@@ -1,14 +1,356 @@
+                continue
+
+            tb_update = self._task_bundle_upgrades.get(task_bundle_upgrade.current_bundle)
+            if tb_update is None:
+                self._task_bundle_upgrades[task_bundle_upgrade.current_bundle] = task_bundle_upgrade
+                tb_update = task_bundle_upgrade
+
+            pf = self._package_file_updates.get(package_file.file_path)
+            if pf is None:
+                self._package_file_updates[package_file.file_path] = package_file
+                pf = package_file
+            pf.task_bundle_upgrades.append(tb_update)
+
+    @staticmethod
+    def _resolve_migrations_for_an_upgrade(
+        task_bundle_upgrade: TaskBundleUpgrade,
+    ) -> Generator[TaskBundleMigration, Any, None]:
+        upgrades_range = determine_task_bundle_upgrades_range(task_bundle_upgrade)
+        for tag_info in upgrades_range:
+            c = Container(task_bundle_upgrade.dep_name)
+            c.tag = tag_info.name
+            c.digest = tag_info.manifest_digest
+            uri_with_tag = c.uri_with_tag
+            script_content = fetch_migration_file(
+                task_bundle_upgrade.dep_name, tag_info.manifest_digest
+            )
+            if script_content:
+                logger.info("Task bundle %s has migration.", uri_with_tag)
+                yield TaskBundleMigration(task_bundle=uri_with_tag, migration_script=script_content)
+            else:
+                logger.info("Task bundle %s does not have migration.", uri_with_tag)
+
+    def resolve_migrations(self) -> None:
+        """Resolve migrations for given task bundle upgrades"""
+        for tb_upgrade in self._task_bundle_upgrades.values():
+            for tb_migration in self._resolve_migrations_for_an_upgrade(tb_upgrade):
+                tb_upgrade.migrations.append(tb_migration)
+            # Quay.io lists tags from the newest to the oldest one.
+            # Migrations must be applied in the reverse order.
+            tb_upgrade.migrations.reverse()
+
+    @staticmethod
+    def _apply_migration(pipeline_file: FilePath, migration: TaskBundleMigration) -> None:
+        if not os.path.exists(pipeline_file):
+            raise ValueError(f"Pipeline file does not exist: {pipeline_file}")
+
+        logger.info(
+            "Apply migration of task bundle %s in package file %s",
+            migration.task_bundle,
+            pipeline_file,
+        )
+
+        fd, migration_file = tempfile.mkstemp()
+        try:
+            os.write(fd, migration.migration_script.encode("utf-8"))
+        finally:
+            os.close(fd)
+
+        with resolve_pipeline(pipeline_file) as file_path:
+            logger.info("Executing migration script %s on %s", migration_file, file_path)
+            cmd = ["bash", migration_file, file_path]
+            logger.debug("Run: %r", cmd)
+            try:
+                proc = sp.run(cmd, stderr=sp.STDOUT, stdout=sp.PIPE)
+                logger.debug("%r", proc.stdout)
```

Comment:
I noticed that this logs a bytestring, not a normal string

Reply:
Yes, it's a byte string. Just in logs, they are not inspected for the most time if there is no failure.

Reply:
Ok, we can keep it that way